### ISSUE TITLE
SAMY? WHIRLPOOL?

### ISSUE THREAD

**Issue by benlaurie** [FellowTraveler/Open-Transactions-old#23]
Why? Why not stick to known and somewhat trusted hashes?

**Comment by FellowTraveler** [comment]
FYI, I am not a programmer by profession (for years now.) One day a few months back, famous hacker Samy was at my house for a party, and he started asking me about digital cash, because he had heard from one of my employees that I had an interest in the stuff.

For whatever reason, this sparked an obsession in me that resulted in the authoring of Open Transactions. I had Samy over at my place a month later while I was writing the initial stages of the code, and he asked me to use TWO hash algorithms instead of one, and then XOR them together. His reasoning? In case one of the algorithms is ever broken, the other one would still be secure and a new one could be swapped in.

So the SAMY hash is actually not some new algorithm -- it's just the XOR of SHA-512 output with WHIRLPOOL output. I chose sha512 and whrlpool based on 2 factors:  (1) a bunch of google research I had done that told me those were secure and respected algorithms. (2) they both have the same output size, so convenient to XOR them together.

Being as you probably know your crypto better than me, I'm open to suggestions if you think it should be changed.

**Comment by benlaurie** [comment]
OK, so ... using two hashes is not a novel idea, SSL does it, for example. Concatenating rather than XOR seems easier to defend (rhetorically, that is), though, and lifts the need to have them the same size.

**Comment by FellowTraveler** [comment]
Yes but it also greatly increases the length of the Open Transactions hashed contract ID, which is currently used as a filename and is already pretty long.

Maybe I'll start concatenating two much shorter hashes.
Maybe we'll switch out when SHA-3 is announced.
For Android version I had to switch to SHA512 instead of SAMY since OpenSSL newest version isn't available yet on Android. I tried to port it myself and got mostly there but not all the way  :(
